lib/spack/spack/spec.py

```diff
@@ -786,6 +786,8 @@ class CompilerFlag(str):
     def __new__(cls, value, **kwargs):
         obj = str.__new__(cls, value)
         obj.propagate = kwargs.pop("propagate", False)
+        obj.flag_group = kwargs.pop("flag_group", value)
+        obj.source = kwargs.pop("source", None)
```

Comment:
Yes, but for the meaintime:

* `flag_group` is the group of flags that introduced a given flag: e.g. if you say `cflags=-x -y`, there will be a `-x` flag tracked in the solver, and we want to remember that it came with `-y`, so the `flag_group` keeps track of `-x -y`
* `source` tracks the type of constraint that introduced the flag: it is needed because the same package can introduce flags via multiple mechanisms (so in terms of applying a consistent order, we have to track this info and assign an arbitrary order based on where the flag constraint came from).